modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java

```diff
@@ -0,0 +1,299 @@
+                                        mvTableStorage,
+                                        replicaMeta
+                                ))
+                        );
+            } else {
+                stopBuildingIndexesIfPrimacyLost(primaryReplicaId);
+
+                return completedFuture(null);
+            }
+        });
+    }
+
+    private void tryStartBuildIndexesForNewPrimaryReplica(
+            int catalogVersion,
+            TablePartitionId primaryReplicaId,
+            MvTableStorage mvTableStorage,
+            ReplicaMeta replicaMeta
+    ) {
+        inBusyLock(busyLock, () -> {
+            if (isLeaseExpire(replicaMeta)) {
+                stopBuildingIndexesIfPrimacyLost(primaryReplicaId);
+
+                return;
+            }
+
+            // TODO: IGNITE-20530 We only need to get write-only indexes
+            for (CatalogIndexDescriptor indexDescriptor : catalogService.indexes(catalogVersion)) {
+                if (primaryReplicaId.tableId() == indexDescriptor.tableId()) {
+                    startBuildIndex(primaryReplicaId, indexDescriptor, mvTableStorage);
+                }
+            }
+        });
+    }
+
+    private void tryStartBuildIndex(
+            TablePartitionId primaryReplicaId,
+            CatalogIndexDescriptor indexDescriptor,
+            MvTableStorage mvTableStorage,
+            ReplicaMeta replicaMeta
+    ) {
+        inBusyLock(busyLock, () -> {
+            if (isLeaseExpire(replicaMeta)) {
+                stopBuildingIndexesIfPrimacyLost(primaryReplicaId);
+
+                return;
+            }
+
+            startBuildIndex(primaryReplicaId, indexDescriptor, mvTableStorage);
+        });
+    }
+
+    private void stopBuildingIndexesIfPrimacyLost(TablePartitionId replicaId) {
+        if (primaryReplicaIds.remove(replicaId)) {
+            // Primary replica is no longer current, we need to stop building indexes for it.
+            indexBuilder.stopBuildingIndexes(replicaId.tableId(), replicaId.partitionId());
+        }
+    }
+
+    private CompletableFuture<MvTableStorage> getMvTableStorageFuture(long causalityToken, TablePartitionId replicaId) {
+        return indexManager.getMvTableStorage(causalityToken, replicaId.tableId());
+    }
+
+    private CompletableFuture<ReplicaMeta> awaitPrimaryReplicaFowNow(TablePartitionId replicaId) {
+        return placementDriver
+                .awaitPrimaryReplica(replicaId, clock.now(), AWAIT_PRIMARY_REPLICA_TIMEOUT, SECONDS)
+                .handle((replicaMeta, throwable) -> {
+                    if (throwable != null) {
+                        Throwable unwrapThrowable = ExceptionUtils.unwrapCause(throwable);
+
+                        if (unwrapThrowable instanceof PrimaryReplicaAwaitTimeoutException) {
+                            return awaitPrimaryReplicaFowNow(replicaId);
+                        } else {
+                            throw new CompletionException(unwrapThrowable);
```

Comment:
I haven’t yet understood why this is not consistent and how to fix it, please explain.

Reply:
Dude! I literally explained the problem in the first statement. Why do you throw an exception instead of returning exceptionally competed future?

Reply:
If you are talking about `completeExceptionally`, I tried to do this, then a compilation error happens, because I need to return `CompletableFuture<ReplicaMeta>` and it says that `CompletableFuture<Object>` will be returned.

Reply:
Are you being serious right now? You're saying that `return CompletableFuture.failedFuture(unwrapThrowable);` lead to compilation error?
> If you are talking about completeExceptionally

No, I'm not talking about that method, why would you assume that?

Reply:
fix it